data-prepper-plugins/geoip-processor/src/main/java/org/opensearch/dataprepper/plugins/processor/databaseenrich/GeoIP2DatabaseReader.java

```diff
@@ -0,0 +1,177 @@
+            }
+
+            if (geoIPDatabases.contains(GeoIPDatabase.ASN)) {
+                final Optional<AsnResponse> asnResponse = enterpriseDatabaseReader.tryAsn(inetAddress);
+                asnResponse.ifPresent(response -> processAsnResponse(response, geoData, fields));
+            }
+
+        } catch (final GeoIp2Exception e) {
+            throw new EnrichFailedException("Address not found in database.");
+        } catch (final IOException e) {
+            throw new EnrichFailedException("Failed to close database readers gracefully. It can be due to expired databases");
+        }
+        return geoData;
+    }
+
+    private void processEnterpriseResponse(final EnterpriseResponse enterpriseResponse, final Map<String, Object> geoData, final List<GeoIPField> fields) {
+        final Continent continent = enterpriseResponse.getContinent();
+        final Country country = enterpriseResponse.getCountry();
+        final Country registeredCountry = enterpriseResponse.getRegisteredCountry();
+        final RepresentedCountry representedCountry = enterpriseResponse.getRepresentedCountry();
+
+        final City city = enterpriseResponse.getCity();
+        final Location location = enterpriseResponse.getLocation();
+        final Postal postal = enterpriseResponse.getPostal();
+        final Subdivision mostSpecificSubdivision = enterpriseResponse.getMostSpecificSubdivision();
+        final Subdivision leastSpecificSubdivision = enterpriseResponse.getLeastSpecificSubdivision();
+
+        extractContinentFields(continent, geoData, fields);
+        extractCountryFields(country, geoData, fields, true);
+        extractRegisteredCountryFields(registeredCountry, geoData, fields);
+        extractRepresentedCountryFields(representedCountry, geoData, fields);
+        extractCityFields(city, geoData, fields, true);
+        extractLocationFields(location, geoData, fields);
+        extractPostalFields(postal, geoData, fields, true);
+        extractMostSpecifiedSubdivisionFields(mostSpecificSubdivision, geoData, fields, true);
+        extractLeastSpecifiedSubdivisionFields(leastSpecificSubdivision, geoData, fields, true);
+    }
+
+    private void processAsnResponse(final AsnResponse asnResponse, final Map<String, Object> geoData, final List<GeoIPField> fields) {
+        extractAsnFields(asnResponse, geoData, fields);
+    }
+
+    @Override
+    public boolean isExpired() {
+        final Instant instant = Instant.now();
+        if (enterpriseDatabaseReader == null) {
+            return true;
+        }
+        if (isEnterpriseDatabaseExpired.get()) {
+            return true;
+        }
+        if (enterpriseDatabaseBuildDate.plus(MAX_EXPIRY_DURATION).isBefore(instant)) {
+            isEnterpriseDatabaseExpired.set(true);
+            closeReader();
+        }
+        return isEnterpriseDatabaseExpired.get();
+    }
+
+    @Override
+    public void retain() {
```

Comment:
Consider finding a way to automatically retain - perhaps on creation. This is not necessary and could be a follow-on PR.